Svelte: Add new generic global navigation UI

vovakulikov opened the pull request:
Closes https://github.com/sourcegraph/sourcegraph/issues/61627

This PR adds a new navigation UI layout that is able to render any arbitrary UI from the children's level. This allows us to render different global navigation menu UI based on context 

- Global product nav on the home page
- Search bar on the search result page
- Repository navigation on the repo and blob UI pages


| Page | Screenshot |
| -------- | -------- |
| Search results with search bar | <img width="1230" alt="Screenshot 2024-04-18 at 16 11 23" src="https://github.com/sourcegraph/sourcegraph/assets/18492575/14f453fd-ebdf-43a3-92b5-796b727d5ea8"> |
| Repository page with repo nav | <img width="1230" alt="Screenshot 2024-04-18 at 16 12 03" src="https://github.com/sourcegraph/sourcegraph/assets/18492575/4300e2dc-816c-4d09-a63b-06bdfa4cbc5c"> |
| With opened sidebar nav | <img width="1230" alt="Screenshot 2024-04-18 at 16 12 33" src="https://github.com/sourcegraph/sourcegraph/assets/18492575/5655a7cf-6537-4e24-a157-4bdd51b97b09"> |


## Test plan
- Go to the search home page, it should be plain navigation that we had prior to this PR
- Go to the search result home, check that navigation rendered search box instead of product navigation (product navigation is still available through sidebar navigation)
- Go to the repository page, and check that the main navigation renders the repository menu instead of the product menu 

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles 

Why does it matter? 

These test plans are there to demonstrate that are following industry standards which are important or critical for our customers. 
They might be read by customers or an auditor. There are meant be simple and easy to read. Simply explain what you did to ensure 
your changes are correct!

Here are a non exhaustive list of test plan examples to help you:

- Making changes on a given feature or component: 
  - "Covered by existing tests" or "CI" for the shortest possible plan if there is zero ambiguity
  - "Added new tests" 
  - "Manually tested" (if non trivial, share some output, logs, or screenshot)
- Updating docs: 
  - "previewed locally" 
  - share a screenshot if you want to be thorough
- Updating deps, that would typically fail immediately in CI if incorrect
  - "CI" 
  - "locally tested" 
-->

camdencheek commented:
Looks really nice! Could you rebase this on `main` to pull in the scrolling changes Felix merged? Tried to review, but looks like there are some changes that meaningfully affect this PR

camdencheek commented:
Fine as a followup: could we add more clear hierarchy in the hamburger popover?

![screenshot-2024-04-18_16-30-09@2x](https://github.com/sourcegraph/sourcegraph/assets/12631702/1f486cc1-4406-4bd2-ae66-85039ab95f85)

camdencheek commented on client/web-sveltekit/src/lib/dom.ts:
```diff
@@ -303,9 +305,19 @@ export const classNames: Action<HTMLElement, string | string[]> = (node, classes
+
+    root.appendChild(target)
+
+    console.log('INITIAL CALL', options)
+
     return {
+        update(options) {
+            console.log('UPDATE', options)
```
intentional console.log?

camdencheek commented on client/web-sveltekit/src/lib/navigation/GlobalHeader.svelte:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    .logo img {
+        &:hover {
+            @keyframes spin {
+                50% {
+                    transform: rotate(180deg) scale(1.2);
+                }
+                100% {
+                    transform: rotate(180deg) scale(1);
+                }
+            }
+
+            @media (prefers-reduced-motion: no-preference) {
+                animation: spin 0.5s ease-in-out 1;
+            }
+        }
```
Am I understanding this right that the logo should spin on hover? I can't seeem to get that working in my browser (firefox)

vovakulikov replied:
@camdencheek can you check that your reduce-motion setting is off? I checked it in my Firefox 125.0.1 (64-bit) and it works. Also notice that only Sourcegraph logo in the global header spins and logo in the sidebar menu doesn't

camdencheek replied:
Oh! Sure enough. I had no idea I had that reduce motion enabled. Works great when I turned that off. Thanks for the pointer

camdencheek commented on client/web-sveltekit/src/lib/navigation/GlobalHeader.svelte:
```diff
@@ -0,0 +1,207 @@
+            display: flex;
+            gap: 1rem;
+            padding: 0;
+            margin: -0.5rem 0 -0.5rem 0;
```
Why do we need negative margins here?

vovakulikov replied:
Here is the structure of this element
```html
<header>
  <logo/>
 
  <navigation-container>
     <plain-list-navigation/>
  <navigation-container/>
</header
```

So this element with negative margins (plain-list-navigation) represents a global navigation list in which we have navigation tabs, since `navigation-container` has some paddings (because this element also represents a container for any arbitrary UI, kind of portal) and we want to ensure that this generic content will look good even if the consumer doesn't add paddings themselves. 

The reason we use negative margins because `plain-list-navigation` UI have tabs with border at the bottom for selected active tab and this border should be at the bottom edge visually. 

<img width="550" alt="Screenshot 2024-04-19 at 15 54 06" src="https://github.com/sourcegraph/sourcegraph/assets/18492575/0b6db966-5468-43d0-8f11-d9467c41f41c">

camdencheek commented on client/web-sveltekit/src/lib/navigation/GlobalHeader.svelte:
```diff
@@ -0,0 +1,207 @@
+    // Custom menu with sidebar navigation controls styles
+    .menu-button {
+        border: none;
+        padding: 0.35rem 0.35rem;
```
```suggestion
        padding: 0.35rem;
```

camdencheek commented on client/web-sveltekit/src/lib/navigation/GlobalHeader.svelte:
```diff
@@ -0,0 +1,207 @@
+        display: flex;
+        align-items: center;
+        background-color: transparent;
+        margin-right: -0.5rem;
```
Rather than using negative margins to cancel out the flexbox gap, can we just put the menu button and the logo in their own div so we can control the spacing between them better? Or maybe just move the menu button outside of the flex div and have the remainder of the header flex?

vovakulikov replied:
I understand the concern about the negative margin. Usually, I think this is better to keep the layout div (any generic wrappers) count as minimal as possible. I don't want to confuse anyone with margins as well, so I added a div wrapper for the logo button and logo icon link.

camdencheek commented on client/web-sveltekit/src/lib/navigation/GlobalSidebarNavigation.svelte:
```diff
@@ -0,0 +1,153 @@
+            display: flex;
+            gap: 0.25rem;
+            align-items: center;
+            padding: 0.67rem 0.5rem 0.5rem 0.5rem;
```
Why is the top padding unbalanced here?

vovakulikov replied:
Note that when you open the sidebar menu, the menu-button icon or Sourcegraph Icon don't move a bit (no UI jumps between close/open states); this non-standard padding is exactly for this. But just realised that probably I can ensure the same height and paddings to avoid jumps rather than calc paddings there.

camdencheek commented on client/web-sveltekit/src/lib/search/input/SearchInput.svelte:
```diff
@@ -293,6 +295,12 @@
             }
         }
+
+        &.compat {
+            padding: 0.25rem;
+            margin: -0.25rem;
+            width: calc(100% + 0.5rem);
```
Instead of negating the additional padding with negative margin in header mode, can we instead just only add it in the other modes?

camdencheek replied:
My suggestion was to just not claim that extra space when we're in header mode. With the negative margins, the negative margin is tightly coupled to the size of the padding we add. No big deal though, no need to follow up, thanks for the explanation 🙂

fkling commented on client/web-sveltekit/src/lib/dom.ts:
```diff
@@ -247,13 +247,15 @@ interface ComputeFitAttributes {
 export const computeFit: Action<HTMLElement, void, ComputeFitAttributes> = node => {
     // Holds the cumulative width of all elements up to element i.
     const widths: number[] = [0]
+    const parentRight = node.getBoundingClientRect().left
```
Despite what I side during our call, this is not actually the right side of the parent, it's the left 😅 
I suggest we rename this do something more neutral, `offset` and add a comment explaining the intention behind it.

```suggestion
    // Used to compute the "end" of each child relative to the parent container.
    // This ensures that the logic here still works when the parent node is moved
    // or when it is scrolled inside a scroll container.
    const offset = node.getBoundingClientRect().left
```

fkling commented on client/web-sveltekit/src/lib/dom.ts:
```diff
@@ -247,13 +247,15 @@ interface ComputeFitAttributes {
 
     function compute(): void {
-        const right = node.getBoundingClientRect().right
+        const rootRect = node.getBoundingClientRect()
+        const right = rootRect.right - rootRect.left
```
We can probably just use `width` here.

```suggestion
        const nodeWidth = node.getBoundingClientRect().width
```

fkling commented on client/web-sveltekit/src/lib/dom.ts:
```diff
@@ -303,9 +305,19 @@ export const classNames: Action<HTMLElement, string | string[]> = (node, classes
 /**
  * An action to move the attached element to the end of the document body.
```
This comment needs to be updated to reflect the new behavior of the this action/function.

fkling commented on client/web-sveltekit/src/lib/navigation/GlobalHeader.svelte:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    export const navigationModeStore = writable<NavigationMode | `${NavigationMode}`>(NavigationMode.PlainNavigation)
+    export const navigationExtensionElement: Writable<HTMLElement | null> = writable(null)
```
In theory, anything that imports this store could assign a new node to this, which we obviously do not want.
You could declare the writable store as local and export a read-only version of it:

```ts
import { readonly } form 'svelte/store'

const extensionElement: Writable<HTMLElement | null> = writable(null)
export const navigationExtensionElement = readonly(extensionElement)
```

Also using the store like this means that `GlobalHeader` can only be instantiated once, which I guess is not that surprising. But in this case I would probably keep it inside `src/routes/`, since it's not supposed to be used anywhere else.

vovakulikov replied:
I was thinking about making the element stream read-only, but I thought that having an intermediate stream was too much for a non-existing problem (I have no problem adding it, though). 

> But in this case I would probably keep it inside src/routes/, since it's not supposed to be used anywhere else.

This is what I wanted to discuss at some point, it looks like our routes directories are growing fast and already have a lot of files, sometimes navigation there isn't that straightforward (pages make a lot of sense but components usually do not) so I moved it to the lib but I agree lib is for reusable components? (but should we change it maybe?)

fkling replied:
> but I thought that having an intermediate stream was too much for a non-existing problem

It doesn't create an intermediate stream. It returns a new object with the same `subscribe` method but without mutation methods:

https://sourcegraph.com/github.com/sveltejs/svelte@b91a67ba6845c5d37dbb6f2fda8a0d057f828dab/-/blob/packages/svelte/src/store/index.js?L193-206

> This is what I wanted to discuss at some point, it looks like our routes directories are growing fast and already have a lot of files, sometimes navigation there isn't that straightforward (pages make a lot of sense but components usually do not) so I moved it to the lib but I agree lib is for reusable components? (but should we change it maybe?)

I haven't arrived at a good pattern yet. On one hand I'd prefer it if components that are used by a single page are located where the page is. On the other hand the specific nature of the `src/routes/` directory makes it impossible to have more meaningful code organization there.

I'd be fine if all components lived in `src/lib/` if we follow specific patterns to indicate shared components. For example the `src/lib/search/` directory could have a `shared/` subfolder and only modules/components inside this folder are allowed to be used by other parts of the app.

fkling commented on client/web-sveltekit/src/lib/navigation/GlobalHeader.svelte:
```diff
@@ -0,0 +1,207 @@
+
+<header class="root">
+    {#if isSidebarNavigationOpen}
+        <GlobalSidebarNavigation onClose={() => setSideNavigationState(false)} />
```
Nit: You could just use

```suggestion
        <GlobalSidebarNavigation onClose={() => isSidebarNavigationOpen = false} />
```
and remove the additional function declaration.

fkling commented on client/web-sveltekit/src/lib/navigation/GlobalSidebarNavigation.svelte:
```diff
@@ -0,0 +1,153 @@
+                background-color: var(--secondary-2);
+            }
+
+            :global([data-icon]) {
+                width: 1rem;
+                height: 1rem;
+                color: var(--icon-color);
+            }
```
`Icon` has a `--size` CSS variable that you can just set. Some for `color`. You don't have to use `global` here:

```suggestion
--size: 1rem;
--color: var(--icon-color)
```

We should rename those variables to be specific to the component though, e.g. `--icon-size`.
 
The same goes for all other places where you are targeting the `Icon` component.

vovakulikov replied:
>Icon has a --size CSS variable that you can just set. Some for color. You don't have to use global here:

I don't think I see it in the Icon component (did you confuse it with Avatar component?), I can add it, though, if you think this would be better.

fkling replied:
Whoops, it accepts a size property.

fkling commented on client/web-sveltekit/src/lib/navigation/GlobalSidebarNavigation.svelte:
```diff
@@ -0,0 +1,153 @@
+            gap: 0.25rem;
+            flex-grow: 1;
+
+            li[aria-current='page'] > :global(a) {
+                background-color: var(--secondary-2);
+            }
```
Why do we not style the `li` element instead?

vovakulikov replied:
There is no reason, changed it to li element

fkling commented on client/web-sveltekit/src/lib/navigation/GlobalSidebarNavigation.svelte:
```diff
@@ -0,0 +1,153 @@
+                background-color: var(--secondary-2);
+            }
+
+            :global(a) {
+                display: flex;
+                flex-wrap: wrap;
+                align-items: center;
+                gap: 0.25rem;
+                padding: 0.375rem 0.75rem;
+                font-size: 1rem;
+
+                :global([data-icon]) {
+                    color: var(--icon-color);
+                }
+
+                &:hover {
+                    background-color: var(--secondary-2);
+                }
+            }
```
Is the style for the list version really that different? If it's not possible to make it work with components own style then I'd either

1. Add a `variant` prop to the component
2. Make the component unstyled and expose the stylable elements via `data-*` props.

vovakulikov replied:
A few questions to clarify 
- > Is the style for the list version really that different? 

different from what component?

> Add a variant prop to the component

- to which component do you suggest to add variant?

My point here is we probably should be ok with not sharing everything as much as possible, layout here is relatively simple I'm afraid if we try to reuse and override some parts of list component it would be more complex rather than simple list in place.

fkling replied:
I meant the `MainNavigationLink` component. It seems that the `:global(a)` is supposed to overwrite the styles of that component.

fkling commented on client/web-sveltekit/src/lib/navigation/GlobalSidebarNavigation.svelte:
```diff
@@ -0,0 +1,153 @@
+                }
+            }
+
+            &.sub-list {
```
This could also just be `> ul`.

fkling commented on client/web-sveltekit/src/routes/[...repo=reporev]/+layout.svelte:
```diff
@@ -38,9 +40,10 @@
-    let visibleNavEntries = navEntries.length
-    $: navEntriesToShow = visibleNavEntries === navEntries.length ? navEntries : navEntries.slice(0, visibleNavEntries)
-    $: overflowMenu = visibleNavEntries !== navEntries.length ? navEntries.slice(visibleNavEntries) : []
+    const visibleNavEntries: Writable<number> = writable(navEntries.length)
```
Why did you turn this into a store? It doesn't look like it has to be one.

vovakulikov replied:
Yeah, I previously did some experiments with extracting it in a separate component to generalise the approach but it's too much for this PR so I rolled back but forgot about the visible entries stream, thanks for spotting.

vovakulikov commented:
I fixed most of the comments and rebased the main (the diff got bigger because the main branch wasn't formatted so I did it in this PR) merging this one, let me know if anything big was left.
